### DISCUSSION TITLE
ELKS with no BIOS dependency

### DISCUSSION THREAD

**Discussion by toncho11** [jbruchon/elks#1737] [General]
Some computers did no have IBM BIOS, but did use 8086/8088.
These are for example:
* Sanyo MBC 555
* Apricot F1, F2 or F10

Both the Apricot and Sanyo load a kind of BIOS from a diskette (RAM BIOS) just before loading their custom version of MS-DOS 2.11. ELKS not having a BIOS dependency by using direct floppy and HD drivers opens the door for the eventual porting of ELKS to these computers.  

OK. I have the Sanyo MBC 555 :)

**Comment by ghaerr** [comment]
> I have the Sanyo MBC 555 :)

You do have a stash of some older stuff!!

> ELKS not having a BIOS dependency by using direct floppy and HD drivers opens the door for the eventual porting of ELKS to these computers.

Well... possibly. [This article](https://en.wikipedia.org/wiki/MBC-550) says the MBC-550 floppy format isn't compatible with the IBM PC. I'm not sure whether the direct floppy driver would work without being specially rewritten. Did the MBC use a NEC 765 FDC?

It also seems the MBC-550 video and hardware clock may not be IBM compatible.

I would say that one could probably get the system up on running on ELKS, but not without a lot of special programming, including a new boot loader, FDC driver, video and hardware timer drivers. It would probably be approximately the work of getting the 8018X port that was accomplished after all BIOS requirements were removed from certain ELKS configurations, plus a floppy FDC driver.

**Comment by toncho11** [comment]
One option for the MBC 555 is to boot from serial port or parallel port. It will require custom boot loader from a floppy disk and a protocol to communicate with an external computer (or device) that will act as a server.

It definitely have advantages such as:
- no limit on space required for the boot image
- no floppy disk writing while testing or updating ELKS (except for the initial bootloader)
- will act as hdd as well

Parallel port is faster, but probably harder to implement.

**Comment by toncho11** [comment]
And what about a custom bootloader for the Sanyo MBC-555 that installs an INT 13h BIOS handler in RAM, and have that handler communicate with a parallel port–based "virtual hard disk"? Would this allow ELKS to perform disk I/O using standard INT 13h calls, redirected through the parallel port?

**Reply by ghaerr** [reply]
Yes, this could certainly be done, which would probably allow an unmodified ELKS boot sector, which uses INT 13h, to load the kernel. Another idea would be to somehow add a standard INT 13 conversion layer after the normal Sanyo boot which redirects INT 13 calls to the Sanyo floppy BIOS. Somewhat technical project and dependent on good documentation on the Sanyo's hardware.

**Reply by toncho11** [reply]
We could in the future reverse engineer the Sanyo DOS boot floppy disk. This will tell us what the Sanyo expect at boot and which parts of the BIOS are RAM loaded and how exactly. 

The Sanyo has only 128 kb memory, so the ROM optimized version will be needed.

**Reply by ghaerr** [reply]
If you can get a copy of the first 512 bytes of the Sanyo DOS boot disk onto Linux, you can disassemble it using `hostdisasm` which is the host-based ELKS disassembler in elkscmd/debug. To make it, cd there and type `make hostdisasm`.
